CRM-21754: Duplicate rows in Activity Details report when address fields are displayed

monishdeb opened the pull request:
Overview
----------------------------------------
Steps to replicate: 
1. MySql 5.7 + sql mode = ONLY_FULL_GROUP_BY
2. Select any address field e.g. city, State, Country to display
3. Create an Activity of any type and select 3 contacts as target, assignee and source. Make sure that each contact has different address field(s).
4. Run Activity Details report

Result: 3 duplicate rows due to three different address field values
Expected: 1 row with merged address field values

Before
----------------------------------------
![screen shot 2018-02-08 at 10 14 56 pm](https://user-images.githubusercontent.com/3735621/35985849-95df4c34-0d1d-11e8-8be0-65e5b37f64fe.png)


After
----------------------------------------
![screen shot 2018-02-08 at 10 14 14 pm](https://user-images.githubusercontent.com/3735621/35985860-9ec00870-0d1d-11e8-87a6-d16e30bb2b31.png)


---

 * [CRM-21754: Duplicate rows in Activity Details report when address fields are displayed ](https://issues.civicrm.org/jira/browse/CRM-21754)

monishdeb commented:
ping @eileenmcnaughton @lcdservices @seamuslee001

eileenmcnaughton commented:
is this a regression?

lcdservices commented:
we have this rolled out in production and it resolves the issue.

seamuslee001 commented:
Jenkins re test this please

seamuslee001 commented:
@monishdeb do you think there is any hope for a UT on this at all i would love if we are able to lock this kind of stuff in.  The only comment i would have about this in a way is that its not immediate obvious which of the country belongs to which of the contacts, you can infer that it starts with the source then assignee etc but just feels a bit hard in a way. Brian not sure what you think @lcdservices

eileenmcnaughton commented:
WRT unit tests I think the class is api_v3_ReportTemplate

lcdservices commented:
@seamuslee001 that was my observation as well. Unlike the email and phone which is explicitly tied to each of the three record types, the address fields are joined with all three. That is the underlying cause of the problem when strict grouping is enabled. I don't like the idea of replicating the address fields separately for all three of the record types as that would make the UI too cluttered. Personally I would lean toward having the address fields tied to the target, as I think that's the most natural interpretation. But @monish's patch retains the existing behavior, while addressing the duplication issue, which one could argue is the best solution for now.

eileenmcnaughton commented:
For comparison I think extended reports allows you to select the addresses as 3 separate column options. If we ARE going down the group_concat patch I think there is a cleaner way to implement -in use elsewhere (I will dig) but I am voting for limiting to target

lcdservices commented:
@monishdeb I'm fine with that.

seamuslee001 commented:
Me too

monishdeb commented:
@lcdservices @eileenmcnaughton @seamuslee001 I have updated the PR and also added UT. So now address fields are only limited to target contact(s). Please have a look.

eileenmcnaughton commented:
@monishdeb the unit test is failing (what version linux / mysql are you using? -if you are not deliberately sorting you might need to accomodate either order as 'natural' sort can vary)

monishdeb commented:
@eileenmcnaughton it passed now :)

eileenmcnaughton commented:
Are the changes to alterAddress still required? I feel like the limiting to target should drop the need for a lot of this patch

monishdeb commented:
@eileenmcnaughton yes its required when we have multiple target contact for a activity and each have their respective address field values. Here's the screenshot:
![screen shot 2018-02-12 at 7 23 49 am](https://user-images.githubusercontent.com/3735621/36081509-ec20571e-0fc5-11e8-91de-0e057f3588f8.png)

Also in the UT I have exposed this same behavior for country and state fields

eileenmcnaughton commented:
OK - there is some existing handling for that based on metadata - I'll see if I can find it & see how they fit

colemanw commented:
@eileenmcnaughton are you still working on this?

monishdeb commented:
^ @eileenmcnaughton

seamuslee001 commented:
Jenkins re test this please

eileenmcnaughton commented:
@seamuslee001 so in the sample data is Łąchowski as a surname - I can't even edit a contact record to have that name! DB error

eileenmcnaughton commented:
Ok - not replicable locally but to do with weird name

colemanw commented:
* ([`r-explain`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-explain)) __Pass__
* ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test)) __Pass__
* ([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code)) __Pass__
* ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) __Pass__
* ([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint)) __Pass__
* ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run)) __Pass__
* ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user)) __Pass__
* ([`r-tech`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-tech)) __Pass__
